### ISSUE TITLE
[SPIKE] "kitchen test" with >2 platforms randomly fails as a result of not copying recipes/inspec files

### ISSUE THREAD

**Issue by MrCyrus** [chef/chef-workstation#1404]
## Description
I've a kitchen.yml file with 9 platforms and would love to set concurrency=9 but "kitchen test" always fails with  concurrency>=3. Random errors happen with recipe files or inspec files not getting copied to random spun up VMs. I can run 'kitchen test' on each individual platform and all works OK but running in parallel results in random issues. 

Some of the errors produced are:

>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 3 actions failed.
>>>>>>     Failed to complete #converge action: [An error occurred while reading the Berksfile:

  No such file or directory @ rb_sysopen - chef-infra_dev-08_aws_agilent_net.crt] on default-win-2016
>>>>>>     Failed to complete #verify action: [Could not fetch inspec profile in "test/integration/default".] on default-win-2012r2
>>>>>>     Failed to complete #verify action: [Could not fetch inspec profile in "test/integration/default".] on default-win-2019

or another error:

>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 3 actions failed.
>>>>>>     Failed to complete #converge action: [An error occurred while reading the Berksfile:

  No such file or directory @ rb_sysopen - chef-infra_dev-08_aws_agilent_net.crt] on default-win-2012r2
>>>>>>     Failed to complete #verify action: [Could not fetch inspec profile in "test/integration/default".] on default-win-2019
>>>>>>     Failed to complete #verify action: [No such file or directory - getcwd] on default-win-2016

## Chef Workstation Version

20.7.96

## Platform Version

Redhat 6.10 64-bit

**Comment by tyler-ball** [comment]
@clintoncwolfe Didn't you mention to me recently that you came across an issue with Test Kitchen concurrency? Tagging you so we can get some eyes on this

**Comment by clintoncwolfe** [comment]
This was the issue I was referring to. It was also mentioned in community slack by this user, in a kitchen-vcenter context, but I just experienced at `-c 3` with kitchen-dokken (using the inspec/inspec project main kitchen.yml file) - it failed to properly copy generated gem files in the cookbook `files/` directory to each VM.

So, "confirmed", I'd say.

**Comment by MrCyrus** [comment]
Not sure if this would help any, but on a server running chef-workstation, kitchen always failed to copy files to one or more newly spun up VMs. One of the suggestions on Slack was to not let kitchen do the parallel converging/testing but rather to use tool ‘parallel’ to run single platform runs of kitchen. This failed with ‘parallel’ telling me I’d run out of memory so I got more memory added to the server and as a test I again ran “kitchen --concurrency 10” and was surprised that this time the kitchen test actually ran successfully on most platforms instead of failing on most.

So I’m wondering if part of the reason kitchen with a large concurrency is failing is because internally it is running out of memory, but that error is never displayed/caught/handled and thus the file copying is silently failing.

Server is ec2 and we changed from t3.small to t3.medium which bumped it from 1G RAM to 4G.
I’ll need to do some more experimenting on this to verify things, and might even get yet more memory added to see if I can get kitchen to run 100% successfully all of the time.

So just wanted to mention this here in case this helps figure out what the problem might be

Tait Cyrus | Automation Specialist | Tata Consultancy Services
Work: +1 719-531-4560 | Webex: https://agilent.webex.com/join/tait.cyrus

**Comment by clintoncwolfe** [comment]
Just some notes on this.
* The main action code immediately spawns a thread, one for each count of `-c`, which defaults to one. So, each (transport) must be thread-safe. 
* At least under kitchen-vcenter, I get `getcwd` failures on about 20% of VMs when concurrency is 5 or more. This isn't perfectly reliable, but is reliable at c = 10.
* My current hypothesis is that as some the threads finish, the sandbox cleanup code runs too early, while the last threads are still wanting to copy over files.

**Comment by clintoncwolfe** [comment]
I've started a [conversation](https://chefcommunity.slack.com/archives/C2B6G1WCQ/p1597862862040200) in the community slack, proposing a few options to move forward. I described the problem as being that TK spawns a thread for each instance on a per-action basis, and so each plugin (the pieces that implement the actions) must be thread safe. Some plugins, such as kitchen-inspec, are known NOT to be thread safe. So some mechanism by which we either disable concurrency entirely or partially is needed. 

I proposed a few ideas: 
A. Document the limitations and live with it (do nothing).
B. Add a method to the plugin superclass, `thread_safe?` which by default returns `false`. Modify the thread-spawning loop to check each plugin to check for thread-safety, and run in serial if unsafe.
C. Modify the concurrency paradigm to use internal forking and IPC.
D. Remove the concurrency option altogether on the next major version release.

Option B was preferred by the community. If implemented, all plugins will initially lose concurrency until they are updated to declare that they are thread-safe.  This may be an obstacle, as some plugins have not seen a release in a while.

An alternative - call it "Optimistic B" - would be to have the superclass `thread_safe?` default to `true`, and then we only implement it returning `false` on plugins that we know have a problem.  That would require fewer software releases and less coordination, but we might find a plugin that is not thread safe and is also unmaintained.

Open to other options / suggestions.
